### PR TITLE
Contribute MyST-NB with jupytext

### PR DESCRIPTION
**Pull request by cooperrc**
@melissawm, I have added a short description in the README using Jupytext to convert or pair notebooks to the desired MyST-NB format. 

This PR also includes 2 more gifs for pairing in the classic notebook and jupyterlab. 

We should also update the "file upload" gif to a `.md` upload.

### CODE REVIEW

File: README.md
Comment (melissawm) [inline README.md:38]: I think something is missing here :)

File: README.md
Comment (melissawm) [inline README.md:64]:
```suggestion
The NumPy tutorials are reviewed and executed as
[MyST-NB](https://myst-nb.readthedocs.io/) notebooks. The benefit is that
```

File: README.md
Comment (melissawm) [inline README.md:80]:
```suggestion
Once installed, you can start your `jupyter lab` or `jupyter notebook`
session in the browser. When launching `jupyter lab` it will ask you to rebuild
to include the Jupytext extension. 
```

File: README.md
Comment (melissawm) [inline README.md:82]:
```suggestion
##### Convert the `.ipynb` file to MyST markdown
```

File: README.md
Comment (melissawm) [inline README.md:68]:
```suggestion
content is easier to review in this markdown format. You can use 
[Jupytext](https://jupytext.readthedocs.io/en/latest/index.html) to convert 
or pair your `.ipynb` file to 
[MyST Markdown](https://github.com/mwouts/jupytext/blob/master/docs/formats.md#myst-markdown) 
format.

Install `jupytext` using:
```

File: README.md
Comment (melissawm) [inline README.md:84]:
```suggestion
To convert a file named `notebook.ipynb` to MyST markdown, run the following command:
```

File: README.md
Comment (melissawm) [inline README.md:90]:
```suggestion
You will now have a `notebook.md` file that is in the MyST markdown format. 
```

File: README.md
Comment (melissawm) [inline README.md:92]:
```suggestion
##### Pair the `.ipynb` to a MyST markdown file
```

File: README.md
Comment (melissawm) [inline README.md:118]:
```suggestion
update either the MyST markdown or notebook file:
```

File: README.md
Comment (melissawm) [inline README.md:124]:
```suggestion
> automatically open MyST files as notebooks. In JupyterLab, you can
```

File: README.md
Comment (melissawm) [inline README.md:143]:
```suggestion
Remember to convert your `.ipynb` to `.md` (MyST markdown) before uploading it. 
```

**Review by melissawm — CHANGES_REQUESTED** [review]
Thank you for this, @cooperrc ! I left some suggestions about MyST mainly noting that MyST markdown is the format and MyST-NB is a tool. I wonder if we should add a note about not using MyST directives in the notebook so we avoid the trouble that @a-elhag went through with his first tutorial - can you help here, Al?

**Comment by cooperrc** [comment]
> Thank you for this, @cooperrc ! I left some suggestions about MyST mainly noting that MyST markdown is the format and MyST-NB is a tool. I wonder if we should add a note about not using MyST directives in the notebook so we avoid the trouble that @a-elhag went through with his first tutorial - can you help here, Al?

It might help to update the "Learn to write a NumPy tutorial" with the best practices for pairing, common mark, and other general advice (like using docs/stable links). I think if we keep the README limited to just a few how-to steps it will be easier to reference for a new contributor. 

@melissawm and @a-elhag, would updating [Learn to write a NumPy tutorial](https://github.com/numpy/numpy-tutorials/blob/master/content/tutorial-style-guide.md) fit within this PR? or should we keep all of this advice in the README?

File: README.md
Comment (cooperrc) [inline README.md:39]: @melissawm , I didn't want to delete existing text without a conversation. 

__Why ~~Jupyter~~ MyST-NB Notebooks?__

This would be a good place to discuss why we are using MyST-NB. Once you have Jupyter+Jupytext, the MyST markdown can be a direct replacement for ipynb without outputs. Since we do not review code outputs, it makes sense to avoid .json files that are meant to couple code+output.

File: README.md
Comment (cooperrc) [reply README.md:38]: whoops, I'll correct that in the next commit

File: README.md
Comment (melissawm) [reply README.md:39]: I see. But I still think most users will come from a plain notebook experience, and might not know what MyST NB is. My thinking was that at some point we could add an action to let the user upload a .ipynb file and the action would create the md file using jupytext - the user wouldn't even need to do that if they didn't want to. 

Maybe @rossbar can also chip in here?

Comment (cooperrc) [reply README.md:39]: That's true. That was the same thinking I had about discussing common mark vs myst-markdown in this Readme. 

I think I should take my own advice and keep this doc simple. 

I do think it would help to discuss these topics in another document. @bjnath's tutorial makes sense to me.

Comment (rossbar) [reply README.md:39]: I agree with keeping it simple - there's a lot of terms and crosstalk (e.g. MyST the markdown flavor, the `myst-parser` which parses MyST-flavored markdown, `MyST-NB` which supports **both** a text-based notebook format *and* MyST-flavored markdown, etc.)

Fortunately none of this stuff should really be relevant for users. My vote would be to minimize trying to disentangle the above mess in our docs and state that `numpy-tutorials` uses a markdown-based format for Jupyter notebooks. Setting up a gh-action to auto-convert ipynb-to-md on push would be straightforward, though it might be more user-friendly to simply do so (with any necessary discussion) as part of the review process. There's a lot of flexibility here so I'm sure we can find an acceptable solution w/ more feedback from authors & reviewers.

Comment (cooperrc) [reply README.md:39]: > Fortunately none of this stuff should really be relevant for users. My vote would be to minimize trying to disentangle the above mess in our docs and state that `numpy-tutorials` uses a markdown-based format for Jupyter notebooks. Setting up a gh-action to auto-convert ipynb-to-md on push would be straightforward, though it might be more user-friendly to simply do so (with any necessary discussion) as part of the review process. 

Yes, it would definitely help to leave out the "disentangling" discussion, especially in this front-facing README. 

I would prefer the notebook pairing process for users (i.e. they have both `.ipynb` and `.md` locally, but submit the `.md`). If they use the jupytext extension, then they don't need a cli approach. If they prefer cli, then the option is still there. In both cases, users can see what is being reviewed and they can update the local ipynb as the review occurs.

**Comment by a-elhag** [comment]
My first time writing a tutorial I thought that since you can go from **Jupyter Notebook -----> Myst Markdown** flawlessly, then it should logically work in the reverse direction too **Myst Markdown -----> Jupyter Notebook**. To add on to what @melissawm and @rossbar said, for the front page I think it would be best to not go into too much detail about the flavour of markdown, at least for the main **README.md**. 

@cooperrc What do you think of making the front page one just for **.ipynb to .md** and/maybe as little details as possible about **.md to ipynb**. In addition to a link to a **How to write NumPy tutorials in the Command Line** document where you can go into more details about the particulars of setting up **Myst Markdown**?

**Comment by melissawm** [comment]
I like the approach of possibly leaving a more advanced workflow for users who wish to deal with that. This could be a separate document or live at the end of the readme (though personally I would prefer a shorter readme)

**Comment by rossbar** [comment]
> I like the approach of possibly leaving a more advanced workflow for users who wish to deal with that. This could be a separate document or live at the end of the readme (though personally I would prefer a shorter readme)

I agree - I don't think we should be making any recommendations about tooling or workflow. There's a ton of flexibility (sphinx-based, nbconvert-based, jupytext-based, vscode plugins, etc.) and I think users need to find what's right for them.

**Comment by cooperrc** [comment]
I'll simplify the steps and remove the MyST-NB/MyST markdown/etc. discussions. I agree with the consensus here: __keep it simple__. 

To @rossbar's point, there are a lot of different workflows that people can use, so its better to leave it up to the user. 

One lingering question though: What is the final format we want to review?
options:
- markdown: this is the default "Download as markdown" option in a Jupyter interface _it does not execute when building the static site_
- MyST markdown: this is the format that `myst-nb` (and also `jupyter-book`) uses to build the static page

If we just want users to submit markdown, then I don't think we need much of a discussion. If we want users to submit MyST markdown, it helps to at least give a potential workflow (and mention that there are plenty of other methods). 

Jupytext plugin recognizes both headers and will allow you to open both `.md` files "as notebooks".

**Comment by rossbar** [comment]
> markdown: this is the default "Download as markdown" option in a Jupyter interface it does not execute when building the static site

Could you clarify what you mean here? We currently have the `myst-nb` extension configured to grab any file with the `.md` extension for conversion to html during the build process. Whether or not anything is executed is a question of whether there are `{code-cell}` directives present in the `.md` document. In other words, any `.md` file in the `content/` folder is included in the sphinx build process.

**Comment by cooperrc** [comment]
> > markdown: this is the default "Download as markdown" option in a Jupyter interface it does not execute when building the static site
> 
> Could you clarify what you mean here? We currently have the `myst-nb` extension configured to grab any file with the `.md` extension for conversion to html during the build process. Whether or not anything is executed is a question of whether there are `{code-cell}` directives present in the `.md` document. In other words, any `.md` file in the `content/` folder is included in the sphinx build process.

Yes, that was not clear. 

In the markdown format, code cells are formatted as

````
```python
import numpy as np
```
````

These cells do not get executed by `myst-nb`. 

The myst markdown uses

````
```{code-cell} ipython3
import numpy as np
```
````

**Comment by rossbar** [comment]
Gotcha, thanks for the clarification. For `jupytext` users, code cells in `.ipynb` format should always map to `{code-cell}` directive blocks automatically (instances of this failing should probably be raised in an issue to `jupytext`). For users writing in markdown directly, it relies on them understanding that the `{code-cell}` is not equivalent to the normal "fenced" code block, which does syntax highlighting but is not executed.

I would expect this not to be too much of a problem as `jupytext` handles it for authors who submit `.ipynb` files, and presumably those submitting markdown files directly are familiar with MyST and aware of the distinction.

**Comment by melissawm** [comment]
Ok. How about this:
- We will review code in myst markdown, with the `{code-cell}` directives;
- If the user wants to, they can install jupytext and "download to myst markdown" to submit their PR.
- If the user does not want to install jupytext, they can upload the ipynb and we'll do the conversion ourselves before starting the review.

The readme can mention uploading ipynb files and include a short optional step ("If you have jupytext installed, you can download your notebook as myst markdown and submit that in your PR"). 

The only problem with that is that notebooks and .md files live in separate folders so this might lead to user error, but I don't think those would be too much of a problem (I'm not expecting to see hundreds of PR's to this repo :))

**Comment by rossbar** [comment]
Sounds good to me! - I had always envisioned reviewers assuming the burden of converting to myst markdown since it's unlikely that authors would be familiar with it and it should be super easy for authors to submit `.ipynb`. If we *do* start hitting reviewer bandwidth limits, then I'd vote for adding a CI workflow to auto-convert with jupytext!

**Comment by cooperrc** [comment]
> Sounds good to me! - I had always envisioned reviewers assuming the burden of converting to myst markdown since it's unlikely that authors would be familiar with it and it should be super easy for authors to submit `.ipynb`. If we _do_ start hitting reviewer bandwidth limits, then I'd vote for adding a CI workflow to auto-convert with jupytext!

If you fork the repo, would the CI convert for you? The reason I ask is that the CI could do a 
```bash
jupytext --set-formats ipynb,myst notebook.ipynb  # Turn notebook.ipynb into a paired ipynb/myst notebook
jupytext --sync notebook.ipynb                  # Update whichever of notebook.ipynb/notebook.md is outdated
```
That way the user/reviewer can update in either format. That's why I had added the discussion of pairing in the README.

**Comment by rossbar** [comment]
Ah, ok - I think I'm finally grasping the context I was missing!

The workflow I've been advocating for assumes that the .ipynb -> myst conversion happens only once (if at all) on PR submission. *I had assumed that all subsequent review/authoring would be done in md format*. This was one of the main motivations for using the md-based notebook format rather than .ipynb (so that the diffs would be sensible, and things like the github suggestion feature work work). So to summarize, this workflow looks something like:
  1. User submits `tutorial.ipynb`
  2. Reviewer runs `jupytext --from notebook --to myst tutorial.ipynb` and pushes to the author's branch.
  3. Review now commences in the "standard" way via github - diffs/suggestions highlight changes and circleCI builds each push, executing the notebooks from the top and showing the rendered output in the `build_artifact`

This workflow makes an assumption that the original authors will be comfortable making edits in the md-based notebook format rather than `.ipynb` files in the browser. Maybe that's too strong an assumption, but the upside is that it avoids having to deal with bi-directional `.ipynb` <-> `.md` conversion at all (leaving users to setup a local workflow that works for them), and ensures that .ipynb files aren't involved in the review process.

**Comment by cooperrc** [comment]
> Ah, ok - I think I'm finally grasping the context I was missing!

:) Probably on me for not being clear. 
 
> The workflow I've been advocating for assumes that the .ipynb -> myst conversion happens only once (if at all) on PR submission. 
>
> 1. User submits `tutorial.ipynb`
> 2. Reviewer runs `jupytext --from notebook --to myst tutorial.ipynb` and pushes to the author's branch.
> 3. Review now commences in the "standard" way via github - diffs/suggestions highlight changes and circleCI builds each push, executing the notebooks from the top and showing the rendered output in the `build_artifact`

This workflow makes sense. The only part is what happens to the `tutorial.ipynb`? We can keep it updated after further commits, but there's not much reason from [Jupytext FAQ](https://jupytext.readthedocs.io/en/latest/faq.html):
> __Which files should I version control?__
>Unless you want to version the outputs, you should version only the text representation. The paired .ipynb file can safely be deleted. It will be recreated locally the next time you > open the notebook (from the text file) and save it.

> Note that if you version both the .md and .ipynb files, you can configure git diff to ignore the diffs on the .ipynb files.

We don't _need_ the `.ipynb`. Is there any risk of confusion if we always review-store-update MyST files and delete `.ipynb`? It just reduces the number of files that need to be organized and kept up to date.

**Comment by rossbar** [comment]
> We don't need the .ipynb. Is there any risk of confusion if we always review-store-update MyST files and delete .ipynb? It just reduces the number of files that need to be organized and kept up to date.

Yup, the original `.ipynb` file is deleted. So to expand on my bulleted list above, step 1 would actually look like:

```bash
jupytext --from notebook --to myst tutorial.ipynb
rm tutorial.ipynb
```

Thus (as soon as we're done with the PRs converting existing material to this workflow, see #44, #47, #49) there will be no `.ipynb` files stored in the `content/` directory, only `.md`

**Comment by cooperrc** [comment]
> Thus (as soon as we're done with the PRs converting existing material to this workflow, see #44, #47, #49) there will be no `.ipynb` files stored in the `content/` directory, only `.md`

Nice, can we remove any failed checks due to outputs still being included? Is it possible to ignore the `.ipynb`'s during circleCI? This would make the submission a little more user-friendly.

**Comment by cooperrc** [comment]
the last commit moved most of these changes to a new tutorial: `content/pairing.md`

I link to the tutorial in the README so a user can find the information if they are so inclined. 

Now, the README has two messages: we review in `.md` and we accept `.ipynb` and myst `.md`.

File: content/pairing.md
Comment (melissawm) [inline content/pairing.md:45]: I would suggest adding this link 

`https://nbformat.readthedocs.io/en/latest/format_description.html`

instead of the json one to this paragraph, might be more useful for less experienced users.

```suggestion
Jupyter notebooks are stored on your disk in a
[json format](https://nbformat.readthedocs.io/en/latest/format_description.html). The json format is
```

File: content/pairing.md
Comment (melissawm) [inline content/pairing.md:36]:
```suggestion
notebooks. Content is easier to review in this markdown format. You can
```

File: content/pairing.md
Comment (melissawm) [inline content/pairing.md:48]: This may be a nit, but it is not necessarily hard to compare json files. The difficulty here is seeing the final rendered notebook for the json file, which is a problem specific to notebooks and not inherent to json files. You may want to reword this, but here's a suggestion:

```suggestion
Python libraries can create. The drawback is that it is hard to see and compare changes made in the notebook file when reviewing pull requests, because this means the reviewers are looking only at the raw json files. 
```

File: content/pairing.md
Comment (melissawm) [inline content/pairing.md:186]:
```suggestion
When you submit a Jupyter notebook to NumPy tutorials, we (the reviewers) will convert
```

File: content/pairing.md
Comment (melissawm) [inline content/pairing.md:247]: Very nit: I think some people may object to your suggestion of vim or emacs as simple text editors :)

File: README.md
Comment (melissawm) [inline README.md:53]:
```suggestion
 * Jupyter notebooks can be launched in [Binder](https://www.mybinder.org), so that users can interact
```

File: README.md
Comment (melissawm) [inline README.md:40]:
```suggestion
to produce static websites and can be opened as notebooks in Jupyter using
[Jupytext](https://jupytext.readthedocs.io). 
```

File: README.md
Comment (melissawm) [inline README.md:43]:
```suggestion
> __Note:__ You should use [CommonMark](https://commonmark.org) markdown
> cells. Jupyter only renders CommonMark.   
```

File: README.md
Comment (melissawm) [reply README.md:53]: Oh! And now in the paragraph above, these are **three** reasons :)

**Review by melissawm — COMMENTED** [review]
Hi @cooperrc - thanks for the patience! Here are my comments, all minor - thanks for doing this, I think this makes it all much clearer, even for me!

File: content/pairing.md
Comment (cooperrc) [reply content/pairing.md:48]: I think that's a nice distinction to make, thanks @melissawm.

File: content/pairing.md
Comment (cooperrc) [reply content/pairing.md:247]: haha! 
simple _text_ editor as opposed to a simple _text editor_. 

This was a conversation I first had in grad school: vim/emacs are editors for simple text i.e. ASCII characters. I would agree, they are not _simple_. 

Maybe cut out the _simple_?

Comment (melissawm) [reply content/pairing.md:246]: I think this is the last change before we merge this; I'd suggest something like *plain text editor*.

**Comment by melissawm** [comment]
Let's merge this and if we think we need to do more, this can be done in a follow-up. Thanks again, @cooperrc !

**Comment by cooperrc** [comment]
> Let's merge this and if we think we need to do more, this can be done in a follow-up. Thanks again, @cooperrc !

Thanks @melissawm! This semester has not had much free time for my FOSS work. Looking forward to getting back to it come May timeframe. 

This merge will definitely help users, we can fine tune it later.

**Pull request merged**
